Src/Web/Web.Shared.Net/AzureAppServiceRoleNameFromHostNameHeaderInitializer.cs

```diff
@@ -0,0 +1,173 @@
+
+                if (!string.IsNullOrEmpty(result) && result.EndsWith(this.WebAppSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    result = result.Substring(0, result.Length - this.WebAppSuffix.Length);
+                }
+
+                this.roleName = result;
+            }
+            catch (Exception ex)
+            {
+                WebEventSource.Log.LogAzureAppServiceRoleNameFromHostNameHeaderInitializerWarning(ex.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets suffix of website name. This must be changed when running in non public Azure region.
+        /// Default value (Public Cloud):  ".azurewebsites.net"
+        /// For US Gov Cloud: ".azurewebsites.us"
+        /// For Azure Germany: ".azurewebsites.de".
+        /// </summary>
+        public string WebAppSuffix { get; set; }
+
+        /// <summary>
+        /// Populates RoleName from the request telemetry associated with the http context.
+        /// If RoleName is empty on the request telemetry, it'll be updated as well so that other telemetry
+        /// belonging to the same requests gets it from request telemetry, without having to parse headers again.
+        /// </summary>
+        /// <remarks>
+        /// RoleName is attempted from every incoming request as opposed to doing this periodically. This is
+        /// done to ensure every request (and associated telemetry) gets the correct RoleName during slot swap.
+        /// </remarks>
+        /// <param name="telemetry">The telemetry item for which RoleName is to be set.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+                {
+                    // RoleName is already populated.
+                    return;
+                }
+
+                string roleName = string.Empty;
+                var context = this.ResolvePlatformContext();
+
+                if (context != null)
+                {
+                    var request = this.GetRequestFromContext(context);
+
+                    if (request != null)
+                    {
+                        if (string.IsNullOrEmpty(request.Context.Cloud.RoleName))
+                        {
+                            if (this.TryGetRoleNameFromHeader(context, out roleName))
+                            {
+                                request.Context.Cloud.RoleName = roleName;
+                            }
+                        }
+                        else
+                        {
+                            roleName = request.Context.Cloud.RoleName;
+                        }
+                    }
+                    else
+                    {
+                        if (!this.TryGetRoleNameFromHeader(context, out roleName))
```

Comment:
can we cache it in statitic variable and refresh only 1 time a minute? It is a very tiny fix with the only drawback of being delayed for 1 minute after swap

Reply:
I gave it a lot of though and decided against it. (Dmitry also suggested same)
As mentioned in the comment for this method, it is intentionally read every request and not periodically.  FunctionsHost is also taking the same approach of reading it in every request.
Even few requests (sampling multiplier may make this worse) with incorrect RoleName is bad, especially considering the support cost. (the symptom is just missing telemetry, so customers keep raising support tickets - it may not be obvious to all customers that the root cause is tied to the way app service slot swap works, and that app insights takes some time to detect this.)